Use smart pointers in Path.cpp

rwlbuis opened the pull request:
#### 8b44517102b796486dd05a51dae00970755b5b86
<pre>
Use smart pointers in Path.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=288084">https://bugs.webkit.org/show_bug.cgi?id=288084</a>

Reviewed by Ryosuke Niwa.

Use smart pointers in Path.cpp.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::ensurePlatformPathImpl):
(WebCore::Path::ensureProtectedImpl):
(WebCore::Path::asImpl):
(WebCore::Path::asImpl const):
(WebCore::Path::moveTo):
(WebCore::Path::addLineTo):
(WebCore::Path::addQuadCurveTo):
(WebCore::Path::addBezierCurveTo):
(WebCore::Path::addArcTo):
(WebCore::Path::addArc):
(WebCore::Path::addEllipse):
(WebCore::Path::addEllipseInRect):
(WebCore::Path::addRect):
(WebCore::Path::addRoundedRect):
(WebCore::Path::addContinuousRoundedRect):
(WebCore::Path::closeSubpath):
(WebCore::Path::segmentsIfExists const):
(WebCore::Path::hasSubpaths const):
(WebCore::Path::fastBoundingRect const):
(WebCore::Path::boundingRect const):
* Source/WebCore/platform/graphics/Path.h:

Canonical link: <a href="https://commits.webkit.org/295396@main">https://commits.webkit.org/295396@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5aa43296cea0e1968ece7f20860dab5fc58cfd45

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104947 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24661 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15082 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110163 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55626 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106988 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25064 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33205 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79684 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107953 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19496 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94713 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59991 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19240 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12791 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55004 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88955 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12837 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112648 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32112 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23616 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88763 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32477 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90939 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88392 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22533 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33286 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11066 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27432 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32036 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37406 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31829 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35170 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33388 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -97,28 +97,28 @@ Path::Path(PathSegment&& segment)
 }
 
-PathImpl& Path::setImpl(Ref<PathImpl>&& impl)
+Ref<PathImpl> Path::setImpl(Ref<PathImpl>&& impl)
```
Does the analysis require this? I don't think it does.

geoffreygaren commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -97,28 +97,28 @@ Path::Path(PathSegment&& segment)
 }
 
-PlatformPathImpl& Path::ensurePlatformPathImpl()
+Ref<PlatformPathImpl> Path::ensurePlatformPathImpl()
```
Usually we create a separate function with the "protected" prefix, so that a caller can decide whether to request a Ref or not.

Something like this:
```
Ref<PlatformPathImpl> Path::ensureProtectedPlatformPathImpl()
{
    return ensurePlatformPathImpl();
}
```

geoffreygaren commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -97,28 +97,28 @@ Path::Path(PathSegment&& segment)
 }
 
-PathImpl& Path::ensureImpl()
+Ref<PathImpl> Path::ensureImpl()
```
`Ref<PathImpl> ensureProtectedImpl() { return ensureImpl(); }`

geoffreygaren commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -134,26 +134,26 @@ void Path::ensureImplForTesting()
 }
 
-PathImpl* Path::asImpl()
+std::optional<Ref<PathImpl>> Path::asImpl()
```
Does the analysis require this? I don't think it does.

If you want to return something nullable, usually we use `RefPtr`.

rniwa reviewed (approved):
This was neutral on MotionMark on iOS and macOS.

rwlbuis committed:
Use smart pointers in Path.cpp
https://bugs.webkit.org/show_bug.cgi?id=288084

Reviewed by Ryosuke Niwa.

Use smart pointers in Path.cpp.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::ensurePlatformPathImpl):
(WebCore::Path::ensureProtectedImpl):
(WebCore::Path::asImpl):
(WebCore::Path::asImpl const):
(WebCore::Path::moveTo):
(WebCore::Path::addLineTo):
(WebCore::Path::addQuadCurveTo):
(WebCore::Path::addBezierCurveTo):
(WebCore::Path::addArcTo):
(WebCore::Path::addArc):
(WebCore::Path::addEllipse):
(WebCore::Path::addEllipseInRect):
(WebCore::Path::addRect):
(WebCore::Path::addRoundedRect):
(WebCore::Path::addContinuousRoundedRect):
(WebCore::Path::closeSubpath):
(WebCore::Path::segmentsIfExists const):
(WebCore::Path::hasSubpaths const):
(WebCore::Path::fastBoundingRect const):
(WebCore::Path::boundingRect const):
* Source/WebCore/platform/graphics/Path.h:

Canonical link: https://commits.webkit.org/295396@main
